openedx/tests/xblock_integration/test_done.py

```diff
@@ -0,0 +1,233 @@
+            cls.xblock2 = ItemFactory.create(
+                parent=cls.unit,
+                category='done',
+                display_name='done_1'
+            )
+
+        cls.course_url = reverse(
+            'courseware_section',
+            kwargs={
+                'course_id': unicode(cls.course.id),
+                'chapter': 'Overview',
+                'section': 'Welcome',
+            }
+        )
+
+    def setUp(self):
+        """
+        Create users
+        """
+        super(TestDone, self).setUp()
+        #EventTestMixin.setUp(self, mock.Mock())
+        for idx, student in enumerate(self.STUDENTS):
+            username = "u{}".format(idx)
+            self.create_account(username, student['email'], student['password'])
+            self.activate_user(student['email'])
+
+        self.staff_user = GlobalStaffFactory()
+
+    def get_handler_url(self, handler, xblock_name=None):
+        """
+        Get url for the specified xblock handler
+        """
+        return reverse('xblock_handler', kwargs={
+            'course_id': unicode(self.course.id),
+            'usage_id': unicode(self.course.id.make_usage_key('done', xblock_name)),
+            'handler': handler,
+            'suffix': ''
+        })
+
+    def enroll_student(self, email, password):
+        """
+        Student login and enroll for the course
+        """
+        self.login(email, password)
+        self.enroll(self.course, verify=True)
+
+    def check_ajax(self, block, data, desired_state):
+        """
+        Make an AJAX call to the XBlock, and assert the state is as
+        desired.
+        """
+        url = self.get_handler_url('toggle_button', 'done_' + str(block))
+        resp = self.client.post(url, json.dumps(data), '')
+        resp_data = json.loads(resp.content)
+        self.assertEqual(resp.status_code, 200)
+        self.assertEqual(resp_data, {"state": desired_state})
+        return resp_data
+
+    def test_done(self):
+        """
+        Walk through a few toggles. Make sure the blocks don't mix up
+        state between them, initial state is correct, and final state
+        is correct.
+        """
+        self.enroll_student(self.STUDENTS[0]['email'], self.STUDENTS[0]['password'])
+        # We confirm we don't have errors rendering the student view
+        self.assert_request_status_code(200, self.course_url)
+        # We confirm the block is initially false
+        self.check_ajax(0, {}, False)
+        self.reset_tracker()
+        self.check_ajax(1, {}, False)
+        self.assert_no_events_were_emitted("edx.done.toggled")
+        # We confirm we can toggle state both ways
+        self.reset_tracker()
+        self.check_ajax(0, {'done': True}, True)
+        self.assert_event_emitted('edx.done.toggled', event_fields={"done": True})
+        self.reset_tracker()
+        self.check_ajax(1, {'done': False}, False)
+        self.assert_event_emitted('edx.done.toggled', event_fields={"done": False})
```

Comment:
I added a test for this. Note that this is a new Mixin. Grading events are not analytics events, and go through a different pipeline which is why this is a little more code than might be expected (see https://github.com/edx/edx-platform/blob/master/lms/djangoapps/courseware/module_render.py#L520 for the logic around this).